apps/labs/posts/close-encounter-with-pandas-and-the-jedis-of-open-source.md

```diff
@@ -0,0 +1,246 @@
+was for an empty DataFrame/Series to be returned if the input is empty (and not an error message), but this didn't 
+fix the case for UDFs (user-defined functions).
+
+```python
+    import pandas as pd
+
+    empty_df = pd.DataFrame({"a": [], "b": []})
+    gb = empty_df.groupby("a", group_keys=True)
+    group = getattr(gb, "b")
+	  # Given the UDF
+	  def last_value(gb):
+        return gb.values[-1]
+    
+    # The inconsistency
+	  group.agg("sum") # Returns an empty Series
+	  # But
+	  group.apply(last_value) # Returns IndexError
+```
+This was a regression issue. Further investigation revealed that as at pandas 1.4 this issue did not exist. In 
+other to maintain consistency I needed to know the particular update to blame for this behavior. Aaron suggested I 
+check out git bisect. It uses binary search to find the commit that introduced a bug. Really cool. 🤓
+
+> Understanding the problem is more than half the solution
+
+After a few conversations and making a few suggested fixes, a solution that maintained consistency with the older 
+version and passed all tests for the current version was selected. The fix was easy; if the groups_key equals zero 
+it means the input was empty and if the function being called on the group isn't one of these aggregate functions 
+*("mad", "skew", "sum", "prod")* then the function won't be called. Proper tests were written (revisioned several 
+times to properly convey the fix and cover all known corner cases) and the fix was merged.
+
+### Internal documentation, the journey ahead, and a new hope
+
+The last few weeks of my internship were interrupted by some unforeseen events. An unfortunate encounter with some 
+corrupt law enforcement officers (a daily struggle in my country) and an immediate mandate to return back to school 
+(thrust me straight into exams) after seven months of industrial strike action. Fortunately for me, through the 
+help of my mentors, I got an extension on my internship (more time to focus on writing this blog and closing all 
+open issues). I am working on writing internal documentation for GroupBy. GroupBy is a big part of pandas and this 
+issue proved to be tougher than I thought as internal documentation requires some background knowledge of the 
+thought process adopted at the time when a feature was been implemented (these are usually hinted at with comments 
+in the code but still they aren't always sufficient in providing clarity).
+
+The internship might have come to an end, with me writing this blog, but the journey from Tatooine is only just 
+beginning. I intend to maintain a relationship with my mentors at Quansight and the maintainers at pandas.
+
+It really has been a journey of enlightenment, knowledge, fun, connection, growth, and Force Awakening. 
+
+I have a greater appreciation now for maintainers in Open Source and I hope to become a major maintainer myself someday. *"It's like poetry, they rhythm"*.😉
```

Comment:
Hadn't heard this quote before. It seems it should be "rhyme", not "rhythm". Maybe you can link the quote to the video for context.

```suggestion
I have a greater appreciation now for maintainers in Open Source and I hope to become a major maintainer myself someday. [*"It's like poetry, they rhyme"*](https://www.youtube.com/watch?v=yFqFLo_bYq0).😉
```